src/librustc/hir/map/mod.rs

```diff
@@ -664,7 +664,7 @@ impl<'hir> Map<'hir> {
             match *node {
                 NodeExpr(ref expr) => {
                     match expr.node {
-                        ExprWhile(..) | ExprLoop(..) => true,
+                        ExprWhile(..) | ExprLoop(..) | ExprIf(..) => true,
```

Comment:
Reverted. I was considering removing all suggestions when involving if blocks due to the accuracy problem mentioned earlier.  Given that now I only provide the suggestion if the types are the same, it can be safely reverted.